### PR TITLE
Introduce Protocol Concept to Merlin Deployment

### PR DESCRIPTION
**Pull request by pradithya**
This merge request introduce `protocol` concept to model deployment in Merlin as part of adding UPI (Universal Prediction Interface) support. 

There are 2 possible protocols values currently:
1. (Default) HTTP_JSON --> deploy model as http service.
2. UPI_V1 --> deploy model as grpc service. Even though the name is already UPI_V1, the protocol enforcement is not yet implemented.
`protocol` value affects both version endpoint and model endpoint. 

### Impact to Version Endpoint

- New `protocol` field in the `VersionEndpoint` object
- When `protocol` is `protocol.UpiV1` the corresponding `InferenceService` expect model to expose grpc service at `defaultGRPCPort` = 9000

```
      ports:
        - containerPort: 9000
          name: h2c
          protocol: TCP
```

- Version endpoint's URL will be the hostname of the inference service, without any scheme or path. e.g. `my-model-1.my-project.domain.com`

### Impact to Model Endpoint

- New `protocol` field in the `ModelEndpoint` object
- When `protocol` is `protocol.UpiV1`, the corresponding `VirtualService` doesn't have path rewrite, and simply map the model endpoint to version endpoint host.

### Other Modifications

- I refactored model endpoint implementation to make the structure more align with other "services" implementation.
- Introduction of `CARAML_` environment variable to replace `MERLIN_` for custom transformer and custom model. 

| Environment Variable      |              Description           | To Replace |
| ------------------------- | ---------------------------------- | ---------- |
| CARAML_HTTP_PORT | Custom Model and custom transformer must open http server listening at this port when `protocol` = `HTTP_JSON` | MERLIN_TRANSFORMER_PORT, MERLIN_PREDICTOR_PORT |
| CARAML_GRPC_PORT | Custom Model and custom transformer must open grpc server listening at this port when `protocol` = `UPI_V1` |     N.A.     |
| CARAML_MODEL_NAME | Model name, shall be used when the custom model or custom transformer requires information about model name |    N.A.      |
| CARAML_MODEL_VERSION | Model version, shall be used when the custom model or custom transformer requires information about model version |  N.A.   |
| CARAML_MODEL_FULL_NAME | Model full name in the format of `${CARAML_MODEL_NAME}-${CARAML_MODEL_FULL_NAME}` |   MERLIN_TRANSFORMER_MODEL_NAME, MERLIN_MODEL_NAME |
| CARAML_MODEL_PREDICTOR_HOST | Predictor host to be used from custom transformer  | MERLIN_TRANSFORMER_MODEL_PREDICT_URL |
| CARAML_ARTIFACT_LOCATION | User's artifact will be located at this path for custom model | MERLIN_ARTIFACT_LOCATION |
| CARAML_DISABLE_LIVENESS_PROBE | Flag to disable liveness probe when set to `true` |  MERLIN_DISABLE_LIVENESS_PROBE  |

**Does this PR introduce a user-facing change?**:

```release-note
Add `protocol` to version endpoint and model endpoint
```

**Checklist**

- [x] Added unit test, integration, and/or e2e tests
- [x] Tested locally
- [x] Update Swagger spec if the PR introduce API changes
- [N.A.] Regenerated Golang and Python client if the PR introduce API changes.
- [N.A. ] Updated documentation

Client and documentation update will be done in separate PR.

### CODE REVIEW

File: api/models/version.go [hunk 1]
```diff
@@ -128,9 +128,8 @@ func (v *Version) GetEndpointByEnvironmentName(envName string) (endpoint *Versio
 }
 
 type ModelOption struct {
-	PyFuncImageName       string
-	PyTorchModelClassName string
```
Comment (karzuo) [inline api/models/version.go:132]: why is this removed? was it never used?

File: api/models/version_endpoint.go [hunk 2]
```diff
@@ -109,17 +122,48 @@ func (ve *VersionEndpoint) IsServing() bool {
...
+	}
+
+	if parsedURL.Scheme == "" {
+		veURL := ve.URL
+		veURL = "//" + veURL
+		parsedURL, err = url.Parse(veURL)
+		if err != nil {
+			return ""
+		}
+	}
```
Comment (karzuo) [inline api/models/version_endpoint.go:164]: L130-142 and L152-164 are repeated codes. Should pull out and put into a new function maybe call getParsedURL()

File: api/service/version_endpoint_service_test.go [hunk 3]
```diff
@@ -2114,3 +2148,25 @@ func TestListContainers(t *testing.T) {
...
+			isEqual := proto.Equal(expProtoMsg, protoMsg)
+			if isEqual {
+				visited[i] = true
+			}
+		}
+	}
+	var numOfMatchElements int
+	for _, val := range visited {
+		if val {
+			numOfMatchElements = numOfMatchElements + 1
```
Comment (karzuo) [inline api/service/version_endpoint_service_test.go:2168]: nitpick: why not use ++

File: db-migrations/29_protocol.up.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE version_endpoints ADD COLUMN protocol VARCHAR(32);
+ALTER TABLE model_endpoints ADD COLUMN protocol VARCHAR(32);
```
Comment (tiopramayudi) [inline db-migrations/29_protocol.up.sql:2]: why do we need protocol field in model_endpoints?

File: api/models/version.go (hunk 1 above)
Comment (pradithya) [reply api/models/version.go:132]: Yes, it's no longer used

File: api/service/model_endpoint_service.go [hunk 5]
```diff
@@ -270,64 +271,80 @@ func (s *modelEndpointsService) createVirtualService(model *models.Model, endpoi
...
-func (s *modelEndpointsService) parseVersionEndpointPath(versionEndpoint *models.VersionEndpoint) (string, error) {
-	veURL, err := url.Parse(versionEndpoint.URL)
-	if err != nil {
-		return "", errors.Wrapf(err, "failed to parse version endpoint url")
+		if !versionEndpoint.IsRunning() {
+			return nil, fmt.Errorf("version Endpoint %s is not running, but %s", versionEndpoint.ID, versionEndpoint.Status)
+		}
+
+		// ensure that all version endpoint destination have same protocol
+		if protocolValue != "" && protocolValue != versionEndpoint.Protocol {
```
[the remark this answers is not in the document]
Comment (pradithya) [reply api/service/model_endpoint_service.go:309]: Currently it's not possible since we have only 1 version endpoint for backing 1 model endpoint. This code is just for future case where we support traffic splitting. In that case, all of the version endpoint must have same protocol.

File: db-migrations/29_protocol.up.sql (hunk 4 above)
Comment (pradithya) [reply db-migrations/29_protocol.up.sql:2]: It's necessary to allow consumer of model endpoint to know the protocol supported without having to know its underlying version endpoint. Having the value in DB is just for convenience.

File: api/models/version_endpoint.go (hunk 2 above)
Comment (pradithya) [reply api/models/version_endpoint.go:164]: Updated, good catch!

File: api/service/version_endpoint_service_test.go (hunk 3 above)
Comment (pradithya) [reply api/service/version_endpoint_service_test.go]: Updated

**Review by karzuo — APPROVED** [review]
Good to go for me.

**Comment by pradithya** [comment]
@tiopramayudi @ariefrahmansyah Is it good to merge?

**Comment by tiopramayudi** [comment]
LGTM

**Review by tiopramayudi — APPROVED** [review]
